File: Baekjoon/DFS/1182.py
# https://www.acmicpc.net/problem/1182

import sys
input = sys.stdin.readline

n, s = map(int, input().split())
arr = list(map(int, input().split()))
result = 0

def dfs(idx, sum):
    global result
    if idx >= n:
        return
    sum += arr[idx]
    if sum == s:
        result += 1
    dfs(idx + 1, sum - arr[idx])
    dfs(idx + 1, sum)
    
dfs(0, 0)
print(result)

File: Baekjoon/DFS/1182.cpp
// https://www.acmicpc.net/problem/1182

#include <iostream>
#include <vector>
using namespace std;

int n, s, cnt = 0;
vector<int> v;

void dfs(int idx, int sum) {
    if (idx == n) {
        if (sum == s) cnt++;
        return;
    }
    dfs(idx + 1, sum + v[idx]);
    dfs(idx + 1, sum);
}

int main() {
    cin >> n >> s;
    v.resize(n);
    for (int i = 0; i < n; i++) cin >> v[i];
    dfs(0, 0);
    if (s == 0) cnt--;
    cout << cnt;
    return 0;
}